### ISSUE TITLE
Freeze on Enumerating File Handle Names

### ISSUE THREAD

**Issue by BrutalSimplicity** [adamdriscoll/PoshInternals#7]
Hey man, awesome library, thanks for the effort.

I was trying to use your library in a small utility for renaming directories that have been locked by Visual Studio (not sure why/how this happens), and I ran into an issue when enumerating the handle names for file types. Basically the program would hang when trying to get the handle's property Name. Best I can tell is that it happens on the NtQueryObject call after checking the GrantedAccess value. After some investigation I found that the offending value was 0x1A019F. After adding it to the check it worked like a charm!

Not sure if you are aware what this particular access value represents (from winnt.h it looks like the set bits are WRITE_OWNER, DELETE, FILE_READ_EA, FILE_READ_DATA, and FILE_READ_ATTRIBUTES), but thought I might bring it to your attention. You'll recognize the updated piece of code below.

Again, thanks. This was very useful.

```cs
  // Query the object name
  if (_typeStr != null && GrantedAccess != 0x0012019f && GrantedAccess != 0x00120189 && GrantedAccess != 0x120089 && GrantedAccess != 0x1A019F) // dont query some objects that could get stuck
  {
    int length;
    NtDll.NtQueryObject(handleDuplicate, OBJECT_INFORMATION_CLASS.ObjectNameInformation, IntPtr.Zero, 0, out length);
    IntPtr ptr = IntPtr.Zero;
    try
    {
      ptr = Marshal.AllocHGlobal(length);
      if (NtDll.NtQueryObject(handleDuplicate, OBJECT_INFORMATION_CLASS.ObjectNameInformation, ptr, length, out length) != NtStatus.Success)
        return;
      _name = Marshal.PtrToStringUni((IntPtr)((long)ptr + 2 * IntPtr.Size));
      if (_typeStr == "File" || _typeStr == "Directory")
      {
        _name = GetRegularFileNameFromDevice(_name);
      }
    }
    finally
    {
      Marshal.FreeHGlobal(ptr);
    }
  }
}
```

**Comment by BrutalSimplicity** [comment]
I now can see that it's my anti-malware program that's causing the issue. I suppose that would make sense. Perhaps some mechanism that prevents programs from tampering with it's internals (or maybe a memory-mapped file). Neat!

If you don't have anything to add please feel free to close this issue. Thanks!

**Comment by adamdriscoll** [comment]
Thanks! Looks like your PR fixed the hang as well!

**Comment by BrutalSimplicity** [comment]
I found the suckers below were still the cause of much frustration. Not sure if that's Universal, or just some particular applications that I'm running. I found that once I stopped querying those my application ran smooth as wine.

One problem that smacked me after turning in that PR is that because I'm using the Thread Pool to give each object query it's own thread to run amok in, their is no"managed" way to kill off these zombie threads. And after several executions of this in the same process those zombies start to stink up the Thread Pool an cause execution to hang, and in some cases it's severe. I think it's a generally okay way to go about things, but certainly wouldn't have it running anything mission-critical. Sorry, I know this is closed, but just thought I'd add some additional insight.

The two culprit values:
```cs
GrantedAccess != 0x1A019F && GrantedAccess != 1048576
```
